pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
             raise PacmanRouteInfoAllocationException(
                 "Unable to use ZonedRoutingInfoAllocator please select a "
                 "different allocator as it needs {} + {} bits".format(
-                    source_bits, max_app_keys_bits))
-        if max_partitions > self.MAX_PARTITIONS_SUPPORTED:
-            raise NotImplementedError()
-        return max_app_keys_bits, key_bits_per_app
+                    app_part_bits, self.__n_bits_atoms_and_mac))
 
-    def _simple_allocate(self, max_app_keys_bits, key_bits_map):
+        # Reserve fixed and check it still works
+        self.__check_fixed()
+        app_part_bits = self.__bits_needed(
+            len(self.__atom_bits_per_app_part) + len(self.__fixed_used))
+        if app_part_bits + self.__n_bits_atoms_and_mac > BITS_IN_KEY:
+            raise PacmanRouteInfoAllocationException(
+                "Unable to use ZonedRoutingInfoAllocator please select a "
+                "different allocator as it needs {} + {} bits".format(
+                    app_part_bits, self.__n_bits_atoms_and_mac))
+
+        if not self.__flexible:
+            if app_part_bits + self.__n_bits_machine + self.__n_bits_atoms > \
+                    BITS_IN_KEY:
+                # We know from above test that all will fit if flexible
+                # Reduce the suggested size of n_bits_atoms
+                self.__n_bits_atoms = \
+                    BITS_IN_KEY - app_part_bits - self.__n_bits_machine
+                self.__n_bits_atoms_and_mac = BITS_IN_KEY - app_part_bits
+            else:
+                # Set the size of atoms and machine for biggest of each
+                self.__n_bits_atoms_and_mac = \
+                    self.__n_bits_machine + self.__n_bits_atoms
+
+    def check_flexible_zones(self):
+        if self.__n_bits_atoms_and_mac == \
+                self.__n_bits_machine + self.__n_bits_atoms:
+            return  # felibible same a s none flexible
+
+        raw_app_part_bits = self.__bits_needed(
+            len(self.__atom_bits_per_app_part))
+
+        while self.__n_bits_atoms >= 0:
+            self.__n_bits_atoms_and_mac = \
+                self.__n_bits_machine + self.__n_bits_atoms
+            if self.__n_bits_atoms_and_mac + raw_app_part_bits <= BITS_IN_KEY:
+                self.__check_fixed()
+                self.__check_fixed()
+                app_part_bits = self.__bits_needed(
+                    len(self.__atom_bits_per_app_part) +
+                    len(self.__fixed_used))
+                if app_part_bits + self.__n_bits_atoms_and_mac <= BITS_IN_KEY:
+                    return  # working result found
+            self.__n_bits_atoms -= 1
+
+        raise PacmanRouteInfoAllocationException(
+            "Unable to use ZonedRoutingInfoAllocator please select a "
+            "different allocator")
+
+    def __check_fixed(self):
```

Comment:
The side effect of this method might be more obvious if it returned the set it created and then the assignment could be done outside.  It might then be called something like "get_fixed_used" as it currently doesn't do any checking.

Reply:
renamed to set_fixed_used